Add `UploadedFileInterface` to valid types of `$value` parameter of `MimeType::isValid()`

akrabat opened the pull request:
The `$value` parameter of `MimeType::isValid()` is passed to [`FileInformationTrait::getFileInfo()`](https://github.com/laminas/laminas-validator/blob/2.51.x/src/File/FileInformationTrait.php#L17) which takes `string|array|object`, so `isValid()`'s `$value` should accept the same types.

akrabat committed:
Add object to $value parameter of MimeType::isValid()

The $value parameter is passed to FileInformationTrait::getFileInfo()
which takes string|array|object, so isValid()'s $value should accept the
same types.

Signed-off-by: Rob Allen <rob@akrabat.com>

akrabat committed:
Update ExcludeMimeType::isValid() signature

This needs to match MimeType::isValid().

Signed-off-by: Rob Allen <rob@akrabat.com>

Ocramius commented:
@akrabat should the baseline be reduced, given your explanation? :thinking:

akrabat commented:
> @akrabat should the baseline be reduced, given your explanation? 🤔

@Ocramius I don't think we can as the fallback if `$value` is not a `string`, `array` or instance of `UploadedFileInterface` is:

```php
return $this->getFileBasedFileInfo($value, $hasType, $hasBasename);
```

Within `getFileBasedFileInfo()`, `$value` is treated as a string, and I'm pretty sure any object that has a `__toString()` would work.

Ocramius commented:
I'm pretty sure any object that has a `__toString()` would work.

Perhaps we should limit to `Stringable` then? Will try to catch you today about this 👍

akrabat committed:
Limit (Exclude)MimeType's type of $value

Investigating the code in more detail shows that $value can only be a
string, an array or an instance of UploadedFileInterface. It cannot be
an object that's Stringable as we call basename() which only supports
real strings.

Signed-off-by: Rob Allen <rob@akrabat.com>

Ocramius commented on src/File/ExcludeMimeType.php:
```diff
@@ -36,8 +36,8 @@ class ExcludeMimeType extends MimeType
      *
-     * @param  string|array|object $value Real file to check for mimetype
-     * @param  array               $file  File data from \Laminas\File\Transfer\Transfer (optional)
+     * @param  string|array|UploadedFileInterface $value Real file to check for mimetype
```
You wanted stringable here? This package has no knowledge of PSR-7

akrabat replied:
No. Because UploadedFileInterface is not Stringable.  I should have imported.

This package already knows PSR-7 because of its current use in [FileInformationTrait](https://github.com/laminas/laminas-validator/blob/2.51.x/src/File/FileInformationTrait.php#L6).

akrabat commented:
Thanks @Ocramius.

I've checked the code in a little more detail and it turns out that `$value` can only be a `string`, an `array` or an instance of `UploadedFileInterface`. It cannot be an object that's `Stringable` as we call `basename()` on which only supports native PHP strings. The only other thing to worry about is that `is_array()` is called on `$value` and again `is_array()` only works with native PHP arrays.

Updated to limit to this.

akrabat committed:
Import UploadedFileInterface

Signed-off-by: Rob Allen <rob@akrabat.com>

Ocramius reviewed (approved):
LGTM, thanks @akrabat!
